java/aiservices/openai/src/main/java/com/microsoft/semantickernel/aiservices/openai/chatcompletion/OpenAIChatCompletion.java

```diff
@@ -82,170 +74,140 @@ public Mono<List<ChatMessageContent>> getChatMessageContentsAsync(ChatHistory ch
-        Mono<List<ChatMessageContent>> results =
-            internalChatMessageContentsAsync(options);
-
-        return results.flatMap(list -> {
-            boolean makeSecondCall = false;
-            for (ChatMessageContent messageContent : list) {
-                if (messageContent.getAuthorRole() == AuthorRole.TOOL) {
-                    makeSecondCall = true;
-                    String content = messageContent.getContent();
-                    String id = messageContent.getModelId();
-                    ChatRequestToolMessage toolMessage = new ChatRequestToolMessage(content, id);
-                    chatRequestMessages.add(toolMessage);
-                }
-            }
-            if (makeSecondCall) {
-                return internalChatMessageContentsAsync(options);
-            }
-            return Mono.just(list);
-        });
+            promptExecutionSettings,
+            kernel);
     }
 
     private Mono<List<ChatMessageContent>> internalChatMessageContentsAsync(
-        ChatCompletionsOptions options) {
-
-        return client
-            .getChatCompletions(getModelId(), options)
-            .filter(choices -> choices.getChoices() != null && !choices.getChoices().isEmpty())
-            .map(this::accumulateResponses)
-            .map(responses ->
-                responses.stream()
-                    .map(ChatResponseCollector::toChatMessageContent)
-                    .collect(Collectors.toList())
+        List<ChatRequestMessage> messages,
+        PromptExecutionSettings settings,
+        Kernel kernel) {
+
+        List<FunctionDefinition> functions = new ArrayList<>();
+        if (kernel != null) {
+            kernel.getPlugins().forEach(plugin ->
+                plugin.getFunctions().forEach((name, function) ->
+                    functions.add(OpenAIFunction.toFunctionDefinition(function.getMetadata(), plugin.getName()))
+                )
             );
+        }
 
-    }
-
+        // Create copy to avoid reactor exceptions when updating the chat options messages internally
+        ChatCompletionsOptions options = getCompletionsOptions(this, new ArrayList<>(messages), functions, settings);
 
-    // non-streaming case
-    private List<ChatResponseCollector> accumulateResponses(ChatCompletions chatCompletions) {
-        List<ChatResponseCollector> collectors = new ArrayList<>();
-        chatCompletions
-            .getChoices()
-            .stream()
-            .map(ChatChoice::getMessage)
-            .filter(Objects::nonNull)
-            .map(accumulateResponse(chatCompletions))
-            .forEach(collectors::add);
-        return collectors;
+        return internalChatMessageContentsAsync(
+                kernel,
+                options,
+                Math.min(MAXIMUM_INFLIGHT_AUTO_INVOKES,
+                        settings != null && settings.getToolCallBehavior() != null ? settings.getToolCallBehavior().getMaximumAutoInvokeAttempts() : 0)
```

Comment:
if settings or toolcallbehaviour is null then the maximum will be 0, is this correct?

Reply:
Yes, that's right. If ToolCallBehavior isn't specified only one called is executed.